File: dscribe/utils/geometry.py
# -*- coding: utf-8 -*-
"""Copyright 2019 DScribe developers

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
import numpy as np

import scipy.sparse
from scipy.spatial.distance import cdist

from ase import Atoms


def get_adjacency_matrix(radius, pos1, pos2=None, output_type="coo_matrix"):
    """Calculates a sparse adjacency matrix by only considering distances
    within a certain cutoff. Uses a k-d tree to reach O(n log(N)) time
    complexity.

    Args:
        radius(float): The cutoff radius within which distances are
            calculated. Distances outside this radius are not included.
        pos1(np.ndarray): A list of N-dimensional positions.
        pos2(np.ndarray): A list of N-dimensional positions. If not provided,
            is assumed to be the same as pos1.
        output_type(str): Which container to use for output data. Options:
            "dok_matrix", "coo_matrix", "dict", or "ndarray". Default:
            "dok_matrix".

    Returns:
        dok_matrix | np.array | coo_matrix | dict: Symmetric sparse 2D
        matrix containing the pairwise distances.
    """
    tree1 = scipy.spatial.cKDTree(
        pos1,
        leafsize=16,
        compact_nodes=True,
        copy_data=False,
        balanced_tree=True,
        boxsize=None
    )
    if pos2 is None:
        pos2 = pos1
    tree2 = scipy.spatial.cKDTree(
        pos2,
        leafsize=16,
        compact_nodes=True,
        copy_data=False,
        balanced_tree=True,
        boxsize=None
    )
    dmat = tree1.sparse_distance_matrix(tree2, radius, output_type=output_type)

    return dmat


def get_adjacency_list(adjacency_matrix):
    """Used to transform an adjacency matrix into an adjacency list. The
    adjacency list provides much faster access to the neighbours of a node.

    Args:
        adjacency_matrix(scipy.sparse.spmatrix): The adjacency matrix from
            which the adjacency list is constructed from. Any of the scipy
            sparse matrix classes.

    Returns:
        list: A list of neighbouring indices. The list of neighbouring indices
        for atom at index i is given by accessing the ith element of this list.
    """
    # Ensure that we have a coo-matrix
    if type(adjacency_matrix) != scipy.sparse.coo_matrix:
        adjacency_matrix = adjacency_matrix.tocoo()

    # Build adjacency list
    adjacency_list = [[] for i in range(adjacency_matrix.shape[0])]
    for i, j in zip(adjacency_matrix.row, adjacency_matrix.col):
        adjacency_list[i].append(j)

    return adjacency_list


def get_extended_system(system, radial_cutoff, centers=None, return_cell_indices=False):
    """Used to create a periodically extended system. If centers are not
    specified, simply takes returns the original system multiplied by an
    integer amount of times in each direction to cover the radial cutoff. If
    centers are provided, returns the exact atoms that are within the given
    radial cutoff from the given centers.

    Args:
        original_system (ase.Atoms): The original periodic system to duplicate.
        radial_cutoff (float): The radial cutoff to use in constructing the
            extended system.
        centers (np.ndarray): Array of xyz-coordinates from which the distance
            is calculated. If provided, these centers are used to calculate the
            exact distance and only atoms within the radial cutoff from these
            centers are returned.
        return_cell_indices (boolean): Whether to return an array of cell
            indices for each atom in the extended system.

    Returns:
        ase.Atoms | tuple: If return_cell_indices is False, returns the new
        extended system. Else returns a tuple containing the new extended
        system as the first entry and the index of the periodically repeated
        cell for each atom as the second entry.
    """
    # Determine the upper limit of how many copies we need in each cell vector
    # direction. We take as many copies as needed to reach the radial cutoff.
    # Notice that we need to use vectors that are perpendicular to the cell
    # vectors to ensure that the correct atoms are included for non-cubic cells.
    cell = np.array(system.get_cell())
    a1, a2, a3 = cell[0], cell[1], cell[2]
    b1 = np.cross(a2, a3, axis=0)
    b2 = np.cross(a3, a1, axis=0)
    b3 = np.cross(a1, a2, axis=0)
    p1 = np.dot(a1, b1) / np.dot(b1, b1) * b1  # Projections onto perpendicular vectors
    p2 = np.dot(a2, b2) / np.dot(b2, b2) * b2
    p3 = np.dot(a3, b3) / np.dot(b3, b3) * b3
    xyz_arr = np.linalg.norm(np.array([p1, p2, p3]), axis=1)
    cell_images = np.ceil(radial_cutoff/xyz_arr)
    nx = int(cell_images[0])
    ny = int(cell_images[1])
    nz = int(cell_images[2])
    n_copies_axis = np.array([nx, ny, nz], dtype=int)

    # If no centers are given, and the cell indices are not requested, simply
    # return the multiplied system. This is much faster.
    if centers is None and not return_cell_indices:

        n_atoms = len(system)
        n_rep = np.product(2*n_copies_axis+1)  # Number of repeated copies
        ext_pos = np.tile(system.get_positions(), (n_rep, 1))

        # Calculate the extended system positions so that the original cell
        # stays in place: both in space and in index
        i_curr = 0
        for m0 in np.append(np.arange(0, nx+1), np.arange(-nx, 0)):
            for m1 in np.append(np.arange(0, ny+1), np.arange(-ny, 0)):
                for m2 in np.append(np.arange(0, nz+1), np.arange(-nz, 0)):
                    ext_pos[i_curr:i_curr+n_atoms] += np.dot((m0, m1, m2), cell)
                    i_curr += n_atoms

        ext_symbols = np.tile(system.get_atomic_numbers(), n_rep)
        extended_system = Atoms(
            positions=ext_pos,
            symbols=ext_symbols,
        )

        return extended_system

    # If centers are given and/or cell indices are needed, the process is done
    # one cell at a time to keep track of the cell inded and the distances.
    # This is a bit slower.
    else:
        # We need to specify that the relative positions should not be wrapped.
        # Otherwise the repeated systems may overlap with the positions taken
        # with get_positions()
        relative_pos = np.array(system.get_scaled_positions(wrap=False))
        numbers = np.array(system.numbers)
        cartesian_pos = np.array(system.get_positions())

        # Create copies of the cell but keep track of the atoms in the
        # original cell
        num_extended = []
        pos_extended = []
        num_extended.append(numbers)
        pos_extended.append(cartesian_pos)
        a = np.array([1, 0, 0])
        b = np.array([0, 1, 0])
        c = np.array([0, 0, 1])
        cell_indices = [np.zeros((len(system), 3), dtype=int)]

        for i in range(-n_copies_axis[0], n_copies_axis[0]+1):
            for j in range(-n_copies_axis[1], n_copies_axis[1]+1):
                for k in range(-n_copies_axis[2], n_copies_axis[2]+1):
                    if i == 0 and j == 0 and k == 0:
                        continue

                    # Calculate the positions of the copied atoms and filter
                    # out the atoms that are farther away than the given
                    # cutoff.
                    num_copy = np.array(numbers)
                    pos_copy = np.array(relative_pos)

                    pos_shifted = pos_copy-i*a-j*b-k*c
                    pos_copy_cartesian = np.dot(pos_shifted, cell)

                    # Only distances to the atoms within the interaction limit
                    # are considered.
                    distances = cdist(pos_copy_cartesian, centers)
                    weight_mask = distances < radial_cutoff

                    # Create a boolean mask that says if the atom is within the
                    # range from at least one atom in the original cell
                    valids_mask = np.any(weight_mask, axis=1)

                    if np.any(valids_mask):
                        valid_pos = pos_copy_cartesian[valids_mask]
                        valid_num = num_copy[valids_mask]
                        valid_ind = np.tile(np.array([i, j, k], dtype=int), (len(valid_num), 1))

                        pos_extended.append(valid_pos)
                        num_extended.append(valid_num)
                        cell_indices.append(valid_ind)

        pos_extended = np.concatenate(pos_extended)
        num_extended = np.concatenate(num_extended)
        cell_indices = np.vstack(cell_indices)

        extended_system = Atoms(
            positions=pos_extended,
            numbers=num_extended,
            cell=cell,
            pbc=False
        )

        return extended_system, cell_indices


def cartesian(arrays, out=None):
    """
    Generate a cartesian product of input arrays.

    Args:
        arrays(sequence of arrays): The arrays from which the product is
            created.
        out(ndarray): Array to place the cartesian product in.

    Returns:
        ndarray: 2-D array of shape (M, len(arrays)) containing cartesian
        products formed of input arrays.

    Example:
    --------
    >>> cartesian(([1, 2, 3], [4, 5], [6, 7]))
    array([[1, 4, 6],
           [1, 4, 7],
           [1, 5, 6],
           [1, 5, 7],
           [2, 4, 6],
           [2, 4, 7],
           [2, 5, 6],
           [2, 5, 7],
           [3, 4, 6],
           [3, 4, 7],
           [3, 5, 6],
           [3, 5, 7]])
    """

    arrays = [np.asarray(x) for x in arrays]
    dtype = arrays[0].dtype

    n = np.prod([x.size for x in arrays])
    if out is None:
        out = np.zeros([n, len(arrays)], dtype=dtype)

    m = int(n / arrays[0].size)
    out[:, 0] = np.repeat(arrays[0], m)
    if arrays[1:]:
        cartesian(arrays[1:], out=out[0:m, 1:])
        for j in range(1, arrays[0].size):
            out[j*m:(j+1)*m, 1:] = out[0:m, 1:]
    return out

File: dscribe/ext/geometry.cpp
/*Copyright 2019 DScribe developers

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#include "geometry.h"

namespace py = pybind11;
using namespace std;

inline vector<double> cross(const vector<double>& a, const vector<double>& b) {
    return {a[1]*b[2]-a[2]*b[1], a[2]*b[0]-a[0]*b[2], a[0]*b[1]-a[1]*b[0]};
}

inline double dot(const vector<double>& a, const vector<double>& b) {
    return a[0]*b[0] + a[1]*b[1] + a[2]*b[2];
}

inline double norm(const vector<double>& a) {
    double accum = 0.0;
    for (size_t i = 0; i < a.size(); ++i) {
        accum += a[i] * a[i];
    }
    return sqrt(accum);
};

double get_volume(const py::array_t<double>& cell) {
    // Extract elements from the cell
    auto cell_u = cell.unchecked<2>();
    double a1 = cell_u(0, 0), a2 = cell_u(0, 1), a3 = cell_u(0, 2);
    double b1 = cell_u(1, 0), b2 = cell_u(1, 1), b3 = cell_u(1, 2);
    double c1 = cell_u(2, 0), c2 = cell_u(2, 1), c3 = cell_u(2, 2);

    // Calculate the cross product of b and c
    double crossX = b2 * c3 - b3 * c2;
    double crossY = b3 * c1 - b1 * c3;
    double crossZ = b1 * c2 - b2 * c1;

    // Dot product of a with the cross product of b and c
    double volume = abs(a1 * crossX + a2 * crossY + a3 * crossZ);

    return volume;
}

std::unordered_map<int, int> count_unique(py::array_t<int> &input_array) {
    // Request a buffer descriptor from the NumPy array
    auto buf = input_array.request();

    // Check that the array is indeed 1D
    if (buf.ndim != 1) {
        throw std::runtime_error("Input array must be 1-dimensional");
    }

    // Pointer to the data
    int *ptr = static_cast<int *>(buf.ptr);
    size_t size = buf.shape[0];

    // Use an unordered_map for counting occurrences
    std::unordered_map<int, int> counts;
    counts.reserve(size); // Reserve space to avoid rehashing

    for (size_t i = 0; i < size; ++i) {
        counts[ptr[i]] += 1;
    }

    return counts;
}

System::System(
    py::array_t<double> positions,
    py::array_t<int> atomic_numbers,
    py::array_t<double> cell,
    bool extra
)
    : positions(positions)
    , atomic_numbers(atomic_numbers)
    , cell(cell)
{
    if (!extra) { return; }

    // Create the default set of interactive atoms, which encompasses the whole
    // system
    unordered_set<int> interactive_atoms = unordered_set<int>();
    int n_atoms = atomic_numbers.size();
    for (int i = 0; i < n_atoms; ++i) {
        interactive_atoms.insert(i);
    }
    this->interactive_atoms = interactive_atoms;

    // Create the default cell indices
    py::array_t<int> cell_indices({n_atoms});
    auto cell_indices_mu = cell_indices.mutable_unchecked<1>();
    for (int i = 0; i < n_atoms; ++i) {
        cell_indices_mu(i) = 0;
    }
    this->cell_indices = cell_indices;

    // Create the default indices
    py::array_t<int> indices({int(n_atoms)});
    auto indices_mu = indices.mutable_unchecked<1>();
    for (int i = 0; i < n_atoms; ++i) {
        indices_mu(i) = i;
    }
    this->indices = indices;
}

System::System(
    py::array_t<double> positions,
    py::array_t<int> atomic_numbers,
    py::array_t<double> cell,
    py::array_t<int> indices,
    py::array_t<int> cell_indices,
    unordered_set<int> interactive_atoms
)
    : positions(positions)
    , atomic_numbers(atomic_numbers)
    , cell(cell)
    , indices(indices)
    , cell_indices(cell_indices)
    , interactive_atoms(interactive_atoms)
{
}

System extend_system(
    py::array_t<double> positions,
    py::array_t<int> atomic_numbers,
    py::array_t<double> cell,
    py::array_t<bool> pbc,
    double cutoff)
{
    if (cutoff < 0) {
        throw invalid_argument("Cutoff must be positive.");
    }
    // Determine the upper limit of how many copies we need in each cell vector
    // direction. We take as many copies as needed to reach the radial cutoff.
    // Notice that we need to use vectors that are perpendicular to the cell
    // vectors to ensure that the correct atoms are included for non-cubic
    // cells.
    auto positions_u = positions.unchecked<2>();
    auto atomic_numbers_u = atomic_numbers.unchecked<1>();
    auto cell_u = cell.unchecked<2>();
    auto pbc_u = pbc.unchecked<1>();
    vector<double> a = {cell_u(0, 0), cell_u(0, 1), cell_u(0, 2)};
    vector<double> b = {cell_u(1, 0), cell_u(1, 1), cell_u(1, 2)};
    vector<double> c = {cell_u(2, 0), cell_u(2, 1), cell_u(2, 2)};

    vector<double> p1 = cross(b, c);
    vector<double> p2 = cross(c, a);
    vector<double> p3 = cross(a, b);

    // Projections of basis vectors onto perpendicular vectors.
    double p1_coeff = dot(a, p1) / dot(p1, p1);
    double p2_coeff = dot(b, p2) / dot(p2, p2);
    double p3_coeff = dot(c, p3) / dot(p3, p3);
    for(double &x : p1) { x *= p1_coeff; }
    for(double &x : p2) { x *= p2_coeff; }
    for(double &x : p3) { x *= p3_coeff; }
    vector<vector<double>> vectors = {p1, p2, p3};

    // Figure out how many copies to take per basis vector. Determined by how
    // many perpendicular projections fit into the cutoff distance.
    vector<int> n_copies_axis(3);
    vector<vector<int>> multipliers;
    for (int i=0; i < 3; ++i) {
        if (pbc_u(i)) {
            double length = norm(vectors[i]);
            double factor = cutoff/length;
            int multiplier = (int)ceil(factor);
            n_copies_axis[i] = multiplier;

            // Store multipliers explicitly into a list in an order that keeps the
            // original system in the same place both in space and in index.
            vector<int> multiples;
            for (int j=0; j < multiplier + 1; ++j) {
                multiples.push_back(j);
            }
            for (int j=-multiplier; j < 0; ++j) {
                multiples.push_back(j);
            }
            multipliers.push_back(multiples);
        } else {
            n_copies_axis[i] = 0;
            multipliers.push_back(vector<int>{0});
        }
    }

    // Calculate the extended system properties
    int n_rep = (2*n_copies_axis[0]+1)*(2*n_copies_axis[1]+1)*(2*n_copies_axis[2]+1);
    int n_atoms = atomic_numbers.size();
    py::array_t<double> ext_pos({n_atoms*n_rep, 3});
    py::array_t<int> ext_atomic_numbers({n_atoms*n_rep});
    py::array_t<int> ext_indices({n_atoms*n_rep});
    py::array_t<int> cell_indices({n_atoms*n_rep});
    auto ext_pos_mu = ext_pos.mutable_unchecked<2>();
    auto ext_atomic_numbers_mu = ext_atomic_numbers.mutable_unchecked<1>();
    auto ext_indices_mu = ext_indices.mutable_unchecked<1>();
    auto cell_indices_mu = cell_indices.mutable_unchecked<1>();
    int i_copy = 0;
    int a_limit = multipliers[0].size();
    int b_limit = multipliers[1].size();
    int c_limit = multipliers[2].size();
    for (int i=0; i < a_limit; ++i) {
        int a_multiplier = multipliers[0][i];
        for (int j=0; j < b_limit; ++j) {
            int b_multiplier = multipliers[1][j];
            for (int k=0; k < c_limit; ++k) {
                int c_multiplier = multipliers[2][k];

                // Precalculate the added vector. It will be used many times.
                vector<double> addition(3, 0);
                for (int m=0; m < 3; ++m) {
                    addition[m] += a_multiplier*a[m] + b_multiplier*b[m] + c_multiplier*c[m];
                };

                // Store the positions, atomic numbers and indices
                for (int l=0; l < n_atoms; ++l) {
                    int index = i_copy*n_atoms + l;
                    ext_atomic_numbers_mu(index) = atomic_numbers_u(l);
                    ext_indices_mu(index) = l;
                    for (int m=0; m < 3; ++m) {
                        ext_pos_mu(index, m) = positions_u(l, m) + addition[m];
                    }
                    cell_indices_mu(index) = i_copy;
                }
                ++i_copy;
            }
        }
    }

    // Store original set of interactive atoms
    unordered_set<int> interactive_atoms = unordered_set<int>();
    for (int i = 0; i < n_atoms; ++i) {
        interactive_atoms.insert(i);
    }

    return System(ext_pos, ext_atomic_numbers, cell, ext_indices, cell_indices, interactive_atoms);
}

py::array_t<double> distancesNumpy(py::detail::unchecked_reference<double, 2> &positions_u)
{
    int n_atoms = positions_u.shape(0);
    py::array_t<double> distances({n_atoms, n_atoms});
    auto distances_mu = distances.mutable_unchecked<2>();
    for (int i = 0; i < n_atoms; ++i) {
        for (int j = i; j < n_atoms; ++j) {
            double dx = positions_u(i, 0) - positions_u(j, 0);
            double dy = positions_u(i, 1) - positions_u(j, 1);
            double dz = positions_u(i, 2) - positions_u(j, 2);
            double distance = sqrt(dx*dx + dy*dy + dz*dz);
            distances_mu(i, j) = distance;
            distances_mu(j, i) = distance;
        }
    }
    return distances;
}

MatrixXd distancesEigen(py::detail::unchecked_reference<double, 2> &positions_u)
{
    int n_atoms = positions_u.shape(0);
    MatrixXd distances(n_atoms, n_atoms);
    for (int i = 0; i < n_atoms; ++i) {
        for (int j = i; j < n_atoms; ++j) {
            double dx = positions_u(i, 0) - positions_u(j, 0);
            double dy = positions_u(i, 1) - positions_u(j, 1);
            double dz = positions_u(i, 2) - positions_u(j, 2);
            double distance = sqrt(dx*dx + dy*dy + dz*dz);
            distances(i, j) = distance;
            distances(j, i) = distance;
        }
    }
    return distances;
}